### ISSUE TITLE
[stackdriver] double free or corruption (out)

### ISSUE THREAD

**Issue by centromere** [fluent/fluent-bit#6745]
## Bug Report

**Describe the bug**
Shortly after start, the application segfaults:

```
Fluent Bit v2.0.8
* Copyright (C) 2015-2022 The Fluent Bit Authors
* Fluent Bit is a CNCF sub-project under the umbrella of Fluentd
* https://fluentbit.io

[2023/01/27 04:58:35] [ info] [fluent bit] version=2.0.8, commit=9444fdc5ee, pid=1
[2023/01/27 04:58:35] [ info] [storage] ver=1.4.0, type=memory, sync=normal, checksum=off, max_chunks_up=128
[2023/01/27 04:58:35] [ info] [cmetrics] version=0.5.8
[2023/01/27 04:58:35] [ info] [ctraces ] version=0.2.7
[2023/01/27 04:58:35] [ info] [input:tail:tail.0] initializing
[2023/01/27 04:58:35] [ info] [input:tail:tail.0] storage_strategy='memory' (memory only)
[2023/01/27 04:58:35] [ info] [input:systemd:systemd.1] initializing
[2023/01/27 04:58:35] [ info] [input:systemd:systemd.1] storage_strategy='memory' (memory only)
[2023/01/27 04:58:35] [ info] [input:systemd:systemd.2] initializing
[2023/01/27 04:58:35] [ info] [input:systemd:systemd.2] storage_strategy='memory' (memory only)
[2023/01/27 04:58:35] [ info] [input:systemd:systemd.3] initializing
[2023/01/27 04:58:35] [ info] [input:systemd:systemd.3] storage_strategy='memory' (memory only)
[2023/01/27 04:58:35] [ info] [input:fluentbit_metrics:fluentbit_metrics.4] initializing
[2023/01/27 04:58:35] [ info] [input:fluentbit_metrics:fluentbit_metrics.4] storage_strategy='memory' (memory only)
[2023/01/27 04:58:35] [ info] [filter:kubernetes:kubernetes.0] https=1 host=kubernetes.default.svc port=443
[2023/01/27 04:58:35] [ info] [filter:kubernetes:kubernetes.0]  token updated
[2023/01/27 04:58:35] [ info] [filter:kubernetes:kubernetes.0] local POD info OK
[2023/01/27 04:58:35] [ info] [filter:kubernetes:kubernetes.0] testing connectivity with API server...
[2023/01/27 04:58:35] [ info] [filter:kubernetes:kubernetes.0] connectivity OK
[2023/01/27 04:58:35] [ info] [output:stackdriver:stackdriver.0] metadata_server set to http://metadata.google.internal
[2023/01/27 04:58:45] [error] [net] connection #538 timeout after 10 seconds to: oauth2.googleapis.com:443
[2023/01/27 04:58:52] [ info] [oauth2] HTTP Status=200
[2023/01/27 04:58:52] [ info] [oauth2] access token from 'oauth2.googleapis.com:443' retrieved
[2023/01/27 04:58:52] [ info] [output:stackdriver:stackdriver.0] worker #0 started
[2023/01/27 04:58:52] [ info] [output:stackdriver:stackdriver.1] metadata_server set to http://metadata.google.internal
[2023/01/27 04:58:52] [ info] [oauth2] HTTP Status=200
[2023/01/27 04:58:52] [ info] [oauth2] access token from 'oauth2.googleapis.com:443' retrieved
[2023/01/27 04:58:52] [ info] [output:stackdriver:stackdriver.1] worker #0 started
[2023/01/27 04:58:52] [ info] [output:stackdriver:stackdriver.2] metadata_server set to http://metadata.google.internal
[2023/01/27 04:58:52] [ info] [oauth2] HTTP Status=200
[2023/01/27 04:58:52] [ info] [oauth2] access token from 'oauth2.googleapis.com:443' retrieved
[2023/01/27 04:58:52] [ info] [output:stackdriver:stackdriver.2] worker #0 started
[2023/01/27 04:58:52] [ info] [output:stackdriver:stackdriver.3] metadata_server set to http://metadata.google.internal
[2023/01/27 04:58:52] [ info] [oauth2] HTTP Status=200
[2023/01/27 04:58:52] [ info] [oauth2] access token from 'oauth2.googleapis.com:443' retrieved
[2023/01/27 04:58:52] [ info] [output:stackdriver:stackdriver.3] worker #0 started
[2023/01/27 04:58:52] [ info] [output:prometheus_exporter:prometheus_exporter.4] listening iface=:: tcp_port=8080
[2023/01/27 04:58:52] [ info] [http_server] listen iface=:: tcp_port=2020
[2023/01/27 04:58:52] [ info] [sp] stream processor started
[2023/01/27 04:58:53] [ info] [input:tail:tail.0] inotify_fs_add(): inode=115779 watch_fd=1 name=/var/log/containers/[REDACTED]
[... 185 lines elided ...]
double free or corruption (out)
[2023/01/27 04:58:53] [engine] caught signal (SIGSEGV)
#0  0x7f391c016602      in  ???() at ???:0
#1  0x7f391c06f767      in  ???() at ???:0
#2  0x7f391c076a59      in  ???() at ???:0
#3  0x7f391c078087      in  ???() at ???:0
[2023/01/27 04:58:53] [ warn] [output:stackdriver:stackdriver.3] error
{
  "error": {
    "code": 401,
    "message": "Request had invalid authentication credentials. Expected OAuth 2 access token, login cookie or other valid authentication credential. See https://developers.google.com/identity/sign-in/web/devconsole-project.",
    "status": "UNAUTHENTICATED",
    "details": [
      {
        "@type": "type.googleapis.com/google.rpc.ErrorInfo",
        "reason": "ACCESS_TOKEN_TYPE_UNSUPPORTED",
        "metadata": {
          "service": "logging.googleapis.com",
          "method": "google.logging.v2.LoggingServiceV2.WriteLogEntries"
        }
      }
    ]
  }
}

#4  0x56490edf014d      in  flb_free() at include/fluent-bit/flb_mem.h:120
#5  0x56490edf13e2      in  flb_sds_destroy() at src/flb_sds.c:470
#6  0x56490f169ea2      in  oauth2_cache_set() at plugins/out_stackdriver/stackdriver.c:90
#7  0x56490f16acf9      in  get_google_token() at plugins/out_stackdriver/stackdriver.c:394
#8  0x56490f17071f      in  cb_stackdriver_flush() at plugins/out_stackdriver/stackdriver.c:2450
#9  0x56490ee15267      in  output_pre_cb_flush() at include/fluent-bit/flb_output.h:527
#10 0x56490f851266      in  co_init() at lib/monkey/deps/flb_libco/amd64.c:117
#11 0xffffffffffffffff  in  ???() at ???:0
```

**Your Environment**
* Version used: `2.0.8`
* Configuration:

```
data:
  filter-kubernetes.conf: |
    [FILTER]
        Name                kubernetes
        Match               kube.*
        Kube_URL            https://kubernetes.default.svc:443
        Kube_CA_File        /var/run/secrets/kubernetes.io/serviceaccount/ca.crt
        Kube_Token_File     /var/run/secrets/kubernetes.io/serviceaccount/token
        Kube_Tag_Prefix     kube.var.log.containers.
        Buffer_Size         0
        Merge_Log           On
        Merge_Log_Key       log_processed
        K8S-Logging.Parser  On
        K8S-Logging.Exclude On
        Use_Kubelet         Off
        tls.verify          Off

    [FILTER]
        Name         nest
        Match        kube.*
        Operation    lift
        Nested_under kubernetes
        Add_prefix   kube_nest.

    [FILTER]
        Name   modify
        Match  kube.*
        Remove logtag
        Remove kube_nest.container_name
        Remove kube_nest.pod_name
        Remove kube_nest.namespace_name
        Remove kube_nest.host
        Rename kube_nest.docker_id kube_nest.container_id

    [FILTER]
        Name          nest
        Match         kube.*
        Operation     nest
        Wildcard      kube_nest.*
        Nest_under    kubernetes
        Remove_prefix kube_nest.

    [FILTER]
        Name      modify
        Match     kube.*
        Condition Key_value_equals stream stderr
        Set       severity ERROR
        Remove    stream

    [FILTER]
        Name      modify
        Match     kube.*
        Condition Key_value_equals stream stdout
        Set       severity INFO
        Remove    stream

    [FILTER]
        Name             modify
        Match            k8s_node.${NODE_NAME}
        Remove_regex     ^((?!^TRANSPORT$|^MESSAGE$).)*$
        Rename TRANSPORT transport
        Rename MESSAGE   message

    [FILTER]
        Name      modify
        Match     k8s_node.${NODE_NAME}
        Condition Key_value_equals transport stderr
        Set       severity ERROR
        Remove    transport

    [FILTER]
        Name      modify
        Match     k8s_node.${NODE_NAME}
        Condition Key_value_equals transport stdout
        Set       severity INFO
        Remove    transport

    [FILTER]
        Name         parser
        Match        kube.*
        Parser       json
        Key_Name     message
        Reserve_Data true

    [FILTER]
        Name      modify
        Match     kube.*
        Condition Key_value_equals level WARN
        Set       level WARNING

    [FILTER]
        Name        modify
        Match       kube.*
        Hard_rename level severity
  filter-os.conf: |
    [FILTER]
        Name            modify
        Match           systemd.*.${NODE_NAME}
        Remove_regex    ^((?!^BOOT_ID$|^HOSTNAME$|^MESSAGE$|^PID$).)*$
        Rename BOOT_ID  boot_id
        Rename HOSTNAME hostname
        Rename MESSAGE  message
        Rename PID      pid
  fluent-bit.conf: |
    [SERVICE]
        Flush        1
        Log_Level    info
        Daemon       off
        Parsers_File parsers.conf
        HTTP_Server  On
        HTTP_Listen  ::
        HTTP_Port    2020
        IPv6         On

    @INCLUDE input-kubernetes.conf
    @INCLUDE input-os.conf
    @INCLUDE filter-kubernetes.conf
    @INCLUDE filter-os.conf
    @INCLUDE output-stackdriver.conf
  input-kubernetes.conf: |
    [INPUT]
        Name             tail
        Tag              kube.*
        Path             /var/log/containers/*.log
        Parser           cri
        DB               /var/db/fluent-bit/flb_kube.db
        Mem_Buf_Limit    16MB
        Skip_Long_Lines  On
        Refresh_Interval 10

    [INPUT]
        Name              systemd
        Tag               k8s_node.${NODE_NAME}
        Path              /var/log/journal
        Systemd_Filter    _SYSTEMD_UNIT=kubelet.service
        Strip_Underscores On
        Mem_Buf_limit     16MB
  input-os.conf: |
    [INPUT]
        Name              systemd
        Tag               systemd.kernel.${NODE_NAME}
        Path              /var/log/journal
        Systemd_filter    _TRANSPORT=kernel
        Strip_Underscores On
        Mem_Buf_limit     16MB

    [INPUT]
        Name              systemd
        Tag               systemd.ssh.${NODE_NAME}
        Path              /var/log/journal
        Systemd_filter    _SYSTEMD_UNIT=ssh.service
        Strip_Underscores On
        Mem_Buf_limit     16MB

    [INPUT]
        Name            fluentbit_metrics
        Tag             internal_metrics
        scrape_interval 2
  output-stackdriver.conf: |
    [OUTPUT]
        Name                 stackdriver
        Match                kube.*
        k8s_cluster_name     [REDACTED]
        k8s_cluster_location [REDACTED]
        Resource             k8s_container
        Location             europe-north1
        Workers              1
        tag_prefix           kube.var.log.containers.
        severity_key         severity

    [OUTPUT]
        Name                 stackdriver
        Match                k8s_node.*
        k8s_cluster_name     [REDACTED]
        k8s_cluster_location [REDACTED]
        Resource             k8s_node
        Location             europe-north1
        Workers              1
        severity_key         severity

    [OUTPUT]
        Name                 stackdriver
        Match                systemd.kernel.*
        Namespace            [REDACTED]
        Job                  kernel
        Task_Id              "0"
        Resource             generic_task
        Location             europe-north1
        Workers              1

    [OUTPUT]
        Name                 stackdriver
        Match                systemd.ssh.*
        Namespace            [REDACTED]
        Job                  ssh
        Task_Id              "0"
        Resource             generic_task
        Location             europe-north1
        Workers              1

    [OUTPUT]
        Name  prometheus_exporter
        Match internal_metrics
        Host  ::
        Port  8080
  parsers.conf: |
    [PARSER]
        Name   nginx
        Format regex
        Regex ^(?<remote>[^ ]*) (?<host>[^ ]*) (?<user>[^ ]*) \[(?<time>[^\]]*)\] "(?<method>\S+)(?: +(?<path>[^\"]*?)(?: +\S*)?)?" (?<code>[^ ]*) (?<size>[^ ]*)(?: "(?<referer>[^\"]*)" "(?<agent>[^\"]*)")?$
        Time_Key time
        Time_Format %d/%b/%Y:%H:%M:%S %z

    [PARSER]
        Name   json
        Format json
        Time_Key time
        Time_Format %d/%b/%Y:%H:%M:%S %z

    [PARSER]
        # http://rubular.com/r/tjUt3Awgg4
        Name cri
        Format regex
        Regex ^(?<time>[^ ]+) (?<stream>stdout|stderr) (?<logtag>[^ ]*) (?<message>.*)$
        Time_Key    time
        Time_Format %Y-%m-%dT%H:%M:%S.%L%z

    [PARSER]
        Name        syslog
        Format      regex
        Regex       ^\<(?<pri>[0-9]+)\>(?<time>[^ ]* {1,2}[^ ]* [^ ]*) (?<host>[^ ]*) (?<ident>[a-zA-Z0-9_\/\.\-]*)(?:\[(?<pid>[0-9]+)\])?(?:[^\:]*\:)? *(?<message>.*)$
        Time_Key    time
        Time_Format %b %d %H:%M:%S
```

* Environment name and version (e.g. Kubernetes? What version?): Kubernetes 1.22.8
* Server type and version: Bare metal
* Operating System and version: Ubuntu 20.04

**Comment by RicardoAAD** [comment]
Hello @centromere,
Thanks for bringing this issue to our attention; as was mentioned in https://github.com/fluent/fluent-bit/issues/6691 this issue is no longer present in the master branch and has been solved by PR https://github.com/fluent/fluent-bit/pull/6594. This fix will be released in Fluent-Bit v2.0.9

You can check these articles to test Fluent-Bit v2.0.9 which includes this change, but as this is not an official version yet, is not intended for a production environment.
https://docs.fluentbit.io/manual/installation/sources/download-source-code
https://docs.fluentbit.io/manual/installation/sources/build-and-install

Please note: the master branch will be our next release 2.0.9. You can also test it with an unofficial image https://github.com/fluent/fluent-bit/tree/master/dockerfiles#ghcrio-topology

**Comment by centromere** [comment]
Thank you @RicardoAAD!
